File: api-gateway/src/routes/agent.ts
import { Router } from 'express'
import { agentService } from '../services/agentService'
import { ValidationError } from '../middleware/errorHandler'

const router = Router()

router.get('/conversations', async (req, res, next) => {
  try {
    const conversations = await agentService.getConversations(req.query.status as string | undefined)
    res.json(conversations)
  } catch (error) {
    next(error)
  }
})

router.get('/conversations/:id', async (req, res, next) => {
  try {
    const conversation = await agentService.getConversation(req.params.id)
    res.json(conversation)
  } catch (error) {
    next(error)
  }
})

router.get('/conversations/:id/messages', async (req, res, next) => {
  try {
    const messages = await agentService.getConversationMessages(req.params.id)
    res.json(messages)
  } catch (error) {
    next(error)
  }
})

router.post('/takeover', async (req, res, next) => {
  try {
    const { conversationId, agentId } = req.body
    if (!conversationId || !agentId) {
      throw new ValidationError('conversationId and agentId are required')
    }

    const conversation = await agentService.takeoverConversation(conversationId, agentId)
    res.json(conversation)
  } catch (error) {
    next(error)
  }
})

router.post('/reply', async (req, res, next) => {
  try {
    const { conversationId, agentId, message } = req.body
    if (!conversationId || !agentId || !message) {
      throw new ValidationError('conversationId, agentId and message are required')
    }

    const reply = await agentService.sendReply(conversationId, agentId, message)
    res.status(201).json(reply)
  } catch (error) {
    next(error)
  }
})

router.get('/tickets', async (_req, res, next) => {
  try {
    const tickets = await agentService.getTickets()
    res.json(tickets)
  } catch (error) {
    next(error)
  }
})

router.post('/tickets', async (req, res, next) => {
  try {
    const { conversationId, userId, type, description, priority, assignedTo } = req.body
    if (!conversationId || !userId || !type || !description) {
      throw new ValidationError('conversationId, userId, type and description are required')
    }

    const ticket = await agentService.createTicket({
      conversationId,
      userId,
      type,
      description,
      priority,
      assignedTo
    })
    res.status(201).json(ticket)
  } catch (error) {
    next(error)
  }
})

export default router

File: agent-service/app/api/agent.py
from typing import Optional
from uuid import uuid4

from fastapi import APIRouter, Depends, HTTPException, Query, status
from pydantic import BaseModel
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.models.database import Conversation, Message, get_db
from app.services.ticket_service import create_ticket as create_ticket_record
from app.services.ticket_service import list_tickets, ticket_to_response

router = APIRouter()


class CreateTicketRequest(BaseModel):
    conversation_id: str
    user_id: str
    type: str
    description: str
    priority: str = "medium"
    assigned_to: Optional[str] = None


class TakeoverRequest(BaseModel):
    conversation_id: str
    agent_id: str


class AgentReplyRequest(BaseModel):
    conversation_id: str
    agent_id: str
    message: str


def format_datetime(value) -> Optional[str]:
    return value.isoformat() if value else None


def conversation_to_response(conversation: Conversation) -> dict:
    return {
        "id": conversation.id,
        "userId": conversation.user_id,
        "channel": conversation.channel,
        "status": conversation.status,
        "currentIntent": conversation.current_intent or "",
        "summary": conversation.summary,
        "assignedAgentId": conversation.assigned_agent_id,
        "createdAt": format_datetime(conversation.created_at),
        "updatedAt": format_datetime(conversation.updated_at),
    }


def message_to_response(message: Message) -> dict:
    metadata = message.meta_data or {}
    return {
        "id": message.id,
        "conversationId": message.conversation_id,
        "senderType": message.sender_type,
        "content": message.content,
        "messageType": metadata.get("messageType", "text"),
        "metadata": metadata,
        "timestamp": format_datetime(message.created_at),
    }

@router.get("/conversations")
async def get_conversations(
    status_filter: Optional[str] = Query(default=None, alias="status"),
    db: AsyncSession = Depends(get_db),
):
    """获取会话列表"""
    query = select(Conversation).order_by(Conversation.updated_at.desc())
    if status_filter:
        query = query.where(Conversation.status == status_filter)
    result = await db.execute(query)
    return [conversation_to_response(conversation) for conversation in result.scalars().all()]


@router.get("/conversations/{conversation_id}")
async def get_conversation(conversation_id: str, db: AsyncSession = Depends(get_db)):
    """获取会话详情"""
    conversation = await db.get(Conversation, conversation_id)
    if conversation is None:
        raise HTTPException(status_code=404, detail="Conversation not found")
    return conversation_to_response(conversation)


@router.get("/conversations/{conversation_id}/messages")
async def get_conversation_messages(
    conversation_id: str,
    db: AsyncSession = Depends(get_db),
):
    """获取会话消息"""
    conversation = await db.get(Conversation, conversation_id)
    if conversation is None:
        raise HTTPException(status_code=404, detail="Conversation not found")

    result = await db.execute(
        select(Message)
        .where(Message.conversation_id == conversation_id)
        .order_by(Message.created_at.asc())
    )
    return [message_to_response(message) for message in result.scalars().all()]

@router.post("/takeover")
async def takeover_conversation(request: TakeoverRequest, db: AsyncSession = Depends(get_db)):
    """接管会话"""
    conversation = await db.get(Conversation, request.conversation_id)
    if conversation is None:
        raise HTTPException(status_code=404, detail="Conversation not found")

    conversation.status = "assigned"
    conversation.assigned_agent_id = request.agent_id
    db.add(conversation)
    await db.flush()
    return conversation_to_response(conversation)

@router.post("/reply", status_code=status.HTTP_201_CREATED)
async def send_reply(request: AgentReplyRequest, db: AsyncSession = Depends(get_db)):
    """发送客服回复"""
    conversation = await db.get(Conversation, request.conversation_id)
    if conversation is None:
        raise HTTPException(status_code=404, detail="Conversation not found")

    if conversation.assigned_agent_id and conversation.assigned_agent_id != request.agent_id:
        raise HTTPException(status_code=409, detail="Conversation assigned to another agent")

    conversation.status = "assigned"
    conversation.assigned_agent_id = request.agent_id

    message = Message(
        id=f"msg_{uuid4().hex}",
        conversation_id=request.conversation_id,
        sender_type="agent",
        content=request.message,
        meta_data={
            "messageType": "text",
            "agentId": request.agent_id,
        },
    )
    db.add(conversation)
    db.add(message)
    await db.flush()
    return message_to_response(message)

@router.get("/tickets")
async def get_tickets(db: AsyncSession = Depends(get_db)):
    """获取工单列表"""
    tickets = await list_tickets(db)
    return [ticket_to_response(ticket) for ticket in tickets]

@router.post("/tickets", status_code=status.HTTP_201_CREATED)
async def create_ticket(request: CreateTicketRequest, db: AsyncSession = Depends(get_db)):
    """创建工单"""
    ticket = await create_ticket_record(
        db=db,
        conversation_id=request.conversation_id,
        user_id=request.user_id,
        ticket_type=request.type,
        description=request.description,
        priority=request.priority,
        assigned_to=request.assigned_to,
    )
    return ticket_to_response(ticket)
